lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,156 @@
+/// `regionToPredicate` (except its terminator). `createPredicate` is used to
+/// create the operations that generate the predicate used by vector.predicate.
+/// These operations will be inserted at the beginning of `regionToPredicate`.
+/// The new vector.predicate will be inserted right after the operations
+/// generating the predicate.
+Optional<PredicateOp> mlir::vector_ext::predicateOp(
+    Operation *op, Region *regionToPredicate, OpBuilder &builder,
+    function_ref<Value(OpBuilder &)> createPredicate) {
+  // TODO: Support multi-block regions.
+  if (!regionToPredicate->hasOneBlock())
+    return llvm::None;
+
+  // Compute the range of operations that will be moved within vector.predicate
+  // and the definitions within the range with users outside the range.
+  Block &blockToPredicate = regionToPredicate->front();
+  auto opsToMove = llvm::make_range(blockToPredicate.begin(),
+                                    Block::iterator(blockToPredicate.back()));
+
+  SmallVector<Value, 8> defsWithUsesOutside;
+  SmallVector<Type, 8> defTypes;
+  getDefsWithUsesOutside(opsToMove, defsWithUsesOutside, defTypes);
+
+  // Set the builder insertion point to the beginning of the loop body to insert
+  // the predicate computation and the vector.predicate.
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(&blockToPredicate);
+  Value predicate = createPredicate(builder);
+
+  // Generate the vector.predicate operation and move 'opsToMove' within its
+  // truePredicateRegion. We have to rewire the def-use chain for those
+  // definitions within the range that have external uses. Those uses are
+  // rewired to the results of the vector.predicate.
+  Location loc = op->getLoc();
+  auto vecPredOp = builder.create<PredicateOp>(loc, defTypes, predicate);
+  assert(defsWithUsesOutside.size() == vecPredOp.getNumResults() &&
+         "Expected same size");
+  for (auto &en : llvm::enumerate(defsWithUsesOutside))
+    en.value().replaceAllUsesWith(vecPredOp.getResult(en.index()));
+
+  Operation *truePredTerminator =
+      &vecPredOp.truePredicateRegion().front().back();
+  moveOperationsBefore(opsToMove, truePredTerminator);
+
+  // The existing terminator of TruePredicateRegion doesn't yield any value.
+  // Replace it with a new terminator that returns the definitions with uses
+  // outside that we just moved.
+  if (vecPredOp.getNumResults() > 0) {
+    builder.setInsertionPoint(truePredTerminator);
+    builder.create<vector_ext::YieldOp>(loc, defsWithUsesOutside);
+    truePredTerminator->erase();
+  }
+
+  return vecPredOp;
+}
+
+/// Utility that predicates the body a tiled loop with a vector.predicate
+/// operation. The vectorization factor used for predication is assumed to be
+/// the step of the tiled loop.
+LogicalResult mlir::vector_ext::predicateTiledLoop(TiledLoopOp loopOp) {
+  if (loopOp.lowerBound().size() > 1)
+    // TODO: Support multi-dim tiled loops.
```

Comment:
Please don't go there too deeply without helpers that are completely independent of linalg.tiled_loop and that we later turn into interfaces, the multi-dim part is quite unlikely to survive at the level you'd create this op.

Reply:
multi-dim support should be transparent to the generic `predicateOp` implementation. It should only impact the way we create the predicate (the lambda function below).

Reply:
nice, thanks!